ci/check-nix-tools.nix

```diff
@@ -14,21 +12,68 @@ writeScript "check-nix-tools.sh" ''
 
   set -euo pipefail
 
-  export PATH="${makeBinPath [ stdenv.shellPackage coreutils nixStable git ]}:$PATH"
+  export PATH="${makeBinPath [ stdenv.shellPackage coreutils nixStable git gawk ]}:$PATH"
 
   cd $(git rev-parse --show-toplevel)
 
-  fail_nix_tools_check() {
-    git diff -w --text > /tmp/nix-tools.patch
-    if [ -n "''${BUILDKITE_JOB_ID:-}" ]; then
-      buildkite-agent artifact upload /tmp/nix-tools.patch --job "$BUILDKITE_JOB_ID"
-    fi
-    echo "ERROR: you need to (run ./nix/regenerate.sh or apply the patch in the buildkite artifact) and commit the changes" >&2
-    exit 1
-  }
-
   # The regenerate script is here by convention
   ./nix/regenerate.sh
 
-  git diff -w --text --exit-code || fail_nix_tools_check
+  # The generated files will appear somewhere under ./nix
+  git add -A nix
+
+  # Check if there are changes staged for commit.
+  if git diff-index --cached --quiet HEAD --; then
+    echo "Generated Nix code is up-to-date."
+    exit 0
+  else
+    echo "Committing changes..."
+    commit_message="Regenerate nix"
+
+    # If on a PR branch, search for a previous regen commit to fix up.
+    commit_fixup=""
+    if [ -n "''${BUILDKITE_PULL_REQUEST_BASE_BRANCH:-}" ]; then
+      git fetch -v origin $BUILDKITE_PULL_REQUEST_BASE_BRANCH
+      commit_fixup=$(git log --pretty=oneline --no-decorate origin/$BUILDKITE_PULL_REQUEST_BASE_BRANCH..HEAD | awk "/$commit_message/ { print \$1; }")
+    fi
+
+    # Create the commit
+    export GIT_COMMITTER_NAME="IOHK"
+    export GIT_COMMITTER_EMAIL="devops+nix-tools@iohk.io"
+    export GIT_AUTHOR_NAME="$GIT_COMMITTER_NAME"
+    export GIT_AUTHOR_EMAIL="$GIT_COMMITTER_EMAIL"
+    if [ -n "$commit_fixup" ]; then
+      git commit --no-gpg-sign --fixup "$commit_fixup"
+    else
+      git commit --no-gpg-sign --message "$commit_message"
+    fi
+
+    # If running in Buildkite...
+    if [ -n "''${BUILDKITE_JOB_ID:-}" ]; then
+
+      # Upload the patch as a Buildkite artifact
+      patch="$BUILDKITE_PIPELINE_SLUG-nix-$BUILDKITE_BUILD_NUMBER.patch"
+      git format-patch --stdout -1 HEAD > "$patch"
+      buildkite-agent artifact upload "$patch" --job "$BUILDKITE_JOB_ID"
+
+      # Push the changes back to the pull request
+      if [ -n "''${BUILDKITE_PULL_REQUEST_REPO:-}" ]; then
+        sshkey="/run/keys/buildkite-$BUILDKITE_PIPELINE_SLUG-ssh-private"
+        if [ -e $sshkey ]; then
+          echo "Authenticating using SSH with $sshkey"
+          export GIT_SSH_COMMAND="ssh -i $sshkey -F /dev/null"
+          remote=$(echo $BUILDKITE_PULL_REQUEST_REPO | sed -e 's=^[a-z]*://github.com/=git@github.com:=')
+          git push $remote HEAD:$BUILDKITE_BRANCH
```

Comment:
since there is a possibility of a `--fixup` commit, this line must assume a possible force push or rather a `--force-with-lease` push to avoid overriding user commits pushed between initial fetch and the force push.

Reply:
There is no `--force` here. The push will fail if non-fast-forward. Which will be fine, because Buildkite will re-run this script at the new branch HEAD anyway.